public/app/plugins/datasource/loki/live_streams.test.ts

```diff
@@ -105,6 +106,39 @@ describe('Live Stream Tests', () => {
     subscription.unsubscribe();
     expect(unsubscribed).toBe(true);
   });
+  it('should reconnect when abnormal error', async () => {
+    const abnormalError = new Error('weird error') as any;
+    abnormalError.code = 1006;
+    const logStream = of({
+      streams: [
+        {
+          stream: { filename: '/var/log/sntpc.log', job: 'varlogs' },
+          values: [['1567025440118944705', 'Kittens']],
+        },
+      ],
+      dropped_entries: null,
+    });
+    const errorStream = throwError(abnormalError);
+    let retries = 0;
+    fakeSocket = of({}).pipe(
+      mergeMap(() => {
+        // When subscribed first time, return logStream and errorStream
+        if (retries++ === 0) {
+          return concat(logStream, errorStream);
+        }
+        // When re-subsribed after abnormal error, return just logStream
+        return logStream;
+      })
+    ) as any;
+    const liveStreams = new LiveStreams();
+    await expect(liveStreams.getStream(makeTarget('url_to_match'), 100)).toEmitValuesWith(received => {
+      const dataBefore = received[0];
+      const dataAfter = received[1];
+      expect(dataBefore).not.toBe(undefined);
+      expect(dataAfter).not.toBe(undefined);
```

Comment:
In order to test that our retry logic works as expected, maybe we should check for specific log rows instead of just toBeDefined?

Reply:
Yes, makes sense! Done in https://github.com/grafana/grafana/pull/29438/commits/e5968b923eb1c957d26800357991e0d6ea02ef8f

Reply:
Beautiful ⭐